vpr/src/place/initial_placement.cpp

```diff
@@ -483,19 +490,81 @@ static std::vector<ClusterBlockId> find_centroid_loc(const t_pl_macro& pl_macro,
+            // VTR_LOG("%s ", g_vpr_ctx.atom().nlist.block_name(atom_blk_id).c_str());
+
+            // Accumulate the x, y, layer, and sub_tile for each atom in each
+            // member of the macro. Remove the offset so the centroid would be
+            // where the head macro should be placed to put the members in the
+            // correct place.
+            acc_x += flat_placement_info.blk_x_pos[atom_blk_id] - member.offset.x;
+            acc_y += flat_placement_info.blk_y_pos[atom_blk_id] - member.offset.y;
+            acc_layer += flat_placement_info.blk_layer[atom_blk_id] - member.offset.layer;
+            acc_sub_tile += flat_placement_info.blk_sub_tile[atom_blk_id] - member.offset.sub_tile;
+            acc_weight++;
+        }
+    }
+    if (acc_weight > 0.f) {
+        // NOTE: We add an offset of 0.5 to prevent us from moving to the tile
+        //       below / to the left due to tiny numerical changes (this
+        //       pretends that each atom is in the center of the tile).
+        centroid.x = std::floor((acc_x / acc_weight) + 0.5f);
+        centroid.y = std::floor((acc_y / acc_weight) + 0.5f);
+        centroid.layer = std::floor((acc_layer / acc_weight) + 0.5f);
+        centroid.sub_tile = std::floor((acc_sub_tile / acc_weight) + 0.5f);
+
+        // TODO: Make this a debug print.
+        // VTR_LOG("\n\t(%d, %d, %d, %d)\n", centroid.x, centroid.y, centroid.layer, centroid.sub_tile);
+    }
+}
+
 static bool try_centroid_placement(const t_pl_macro& pl_macro,
                                    const PartitionRegion& pr,
                                    t_logical_block_type_ptr block_type,
                                    e_pad_loc_type pad_loc_type,
                                    vtr::vector<ClusterBlockId, t_block_score>& block_scores,
                                    BlkLocRegistry& blk_loc_registry,
+                                   const FlatPlacementInfo& flat_placement_info,
                                    vtr::RngContainer& rng) {
     auto& block_locs = blk_loc_registry.mutable_block_locs();
 
     t_pl_loc centroid_loc(OPEN, OPEN, OPEN, OPEN);
     std::vector<ClusterBlockId> unplaced_blocks_to_update_their_score;
 
-    unplaced_blocks_to_update_their_score = find_centroid_loc(pl_macro, centroid_loc, blk_loc_registry);
+    if (!flat_placement_info.valid) {
+        // If a flat placement is not provided, use the centroid of connected
+        // blocks which have already been placed.
+        unplaced_blocks_to_update_their_score = find_centroid_loc(pl_macro, centroid_loc, blk_loc_registry);
+    } else {
+        // If a flat placement is provided, use the flat placement to get the
+        // centroid.
+        find_centroid_loc_from_flat_placement(pl_macro, centroid_loc, flat_placement_info);
+        // If a centroid could not be found, or if the tile is not legal
+        // fall-back on the centroid of the neighbor blocks of this block.
```

Comment:
I'd comment that this may be more disruptive than needed to flat placement reconstruction.